Improve Digital Adapter Implementation

agregorio-improve opened the pull request.

ix-certification commented on improvedigitaladserverjsclient.js:
```diff
@@ -0,0 +1,230 @@
+  };
+
+  this.createRequest = function (requestObject, requestParameters, extraRequestParameters) {
+    if (requestParameters.httpRequestType !== this.CONSTANTS.HTTP_REQUEST_TYPE.GET) {
```
If i'm following the code correctly, `httpRequestType` is hard coded to be `CONSTANTS.HTTP_REQUEST_TYPE.GET`. Is there a point of having this additional layer of check?

agregorio-improve replied:
We could extend our client code to support POST so we left the option to specify POST.  We can remove this if you wish.

ix-certification commented on improvedigitaladserverjsclient.js:
```diff
@@ -0,0 +1,230 @@
+    if (requestParameters.httpRequestType !== this.CONSTANTS.HTTP_REQUEST_TYPE.GET) {
+      return this.getErrorReturn(this.CONSTANTS.ERROR_CODES.BAD_HTTP_REQUEST_TYPE_PARAM);
+    }
+    if (!requestParameters.libVersion) {
```
Same concern as the above comment.

agregorio-improve replied:
This will be removed in the next submission.

ix-certification commented on improvedigitaladserverjsclient.js:
```diff
@@ -0,0 +1,230 @@
+    var impressionObject = void 0;
+    var counter = void 0;
+    if (utils.isArray(requestObject)) {
+      for (counter = 0; counter < requestObject.length; counter++) {
```
Since counter is used in other for loops as well, can you please use `for (var counter=0; ...)` instead?

agregorio-improve replied:
This will be fixed in the next submission

ix-certification commented on improvedigitaladserverjsclient.js:
```diff
@@ -0,0 +1,230 @@
+            if (returnIdMappings) {
+              returnObject.idMappings.pop();
+            }
+            counter--;
```
Is there a chance that this causes a infinite loop? Can you please explain what this part (inside for loop) of the code is doing?

agregorio-improve replied:
Our client supports the submission of one impression per request or multiple impressions per request.  For technical reasons, we are not making use of single request in our adapters for now.  This code is only triggered in single request mode so there is no chance of an infinite loop.

Anyway, to explain the logic, if single request is turned on, the following happens:

1) Keep pushing impression objects into the request until it exceeds MAX_URL_LENGTH.
2) If it exceeds MAX_URL_LENGTH and there is more than 1 impression object, pop the last impression object from the request, decrement counter so that you can process it as part of another request.

Not that if a single impression request exceeds MAX_URL_LENGTH (which only applies to older browsers), it still gets sent since we don't apply this check for requests with just one impression in them.

agregorio-improve commented on improve-digital-htb.js:
```diff
@@ -237,42 +275,41 @@ function ImproveDigitalHtb(configs) {
-            if (!curBid) {
-                if (__profile.enabledAnalytics.requestTime) {
+            if (!curBid || typeof curBid.price === 'undefined' || curBid.price === 0) {
+                if (__profile.enabledAnalytics.requestTime && __baseClass._emitStatsEvent) {
```
It is possible to return JSON where price is 0 or undefined.  This should be interpreted as a no bid, hence the check.  I can either leave this check here or move the check to the loop above where we traverse the bids and set curBid.

In terms of the check on __baseClass._emitStatsEvent, if we don't check, the test code blows up because __baseClass._emitStatsEvent is not defined in this stubbed state.  The following error is produced:

TypeError: __baseClass._emitStatsEvent is not a function

agregorio-improve commented on improve-digital-htb.js:
```diff
@@ -237,42 +275,41 @@ function ImproveDigitalHtb(configs) {
             /* No matching bid found so its a pass */
-            if (!curBid) {
-                if (__profile.enabledAnalytics.requestTime) {
+            if (!curBid || typeof curBid.price === 'undefined' || curBid.price === 0) {
```
It is possible to return JSON where price is 0 or undefined. This should be interpreted as a no bid, hence the check. I can either leave this check here or move the check to the loop above where we traverse the bids and set curBid.

Which do you prefer?

agregorio-improve commented on improve-digital-htb.js:
```diff
@@ -421,7 +482,7 @@ function ImproveDigitalHtb(configs) {
                 inputCentsMultiplier: 1, // Input is in cents
                 outputCentsDivisor: 1, // Output as cents
-                outputPrecision: 0, // With 0 decimal places
+                outputPrecision: 2, // With 2 decimal places
```
Will fix this with the next submission.

agregorio-improve commented on utils.js:
```diff
@@ -0,0 +1,33 @@
+'use strict';
```
Will fix in the next submission.

ix-certification replied:
Hi, just following up on this.

agregorio-improve commented on utils.js:
```diff
@@ -0,0 +1,33 @@
+    return Object.prototype.toString.call(object) === '[object ' + _t + ']';
+};
+
+exports.isArray = function (object) {
```
Will fix in the next submission.

agregorio-improve commented on utils.js:
```diff
@@ -0,0 +1,33 @@
+    };
+}();
+
+function _getUniqueIdentifierStr() {
```
Will fix in the next submission.

agregorio-improve commented on improve-digital-htb.js:
```diff
@@ -25,15 +24,16 @@ var Size = require('size.js');
-//? }
+
+
+var improvedigitalclient = require('./improvedigitaladserverjsclient.js');
```
This will be fixed in the next submission.

agregorio-improve commented on improve-digital-htb.js:
```diff
@@ -179,8 +216,9 @@ function ImproveDigitalHtb(configs) {
-        var callbackId = 0;
-        __baseClass._adResponseStore[callbackId] = adResponse;
+        if (adResponse.id) {
+            __baseClass._adResponseStore[id] = adResponse;
```
This will be fixed in the next submission.

Response should always have id set.

agregorio-improve commented on improve-digital-htb.js:
```diff
@@ -151,19 +164,43 @@ function ImproveDigitalHtb(configs) {
+            for (parcelCounter = 0; parcelCounter < returnParcels.length; parcelCounter++) {
+                var parcel = returnParcels[parcelCounter];
+                // Delete the adUnitId
+                parcel.xSlotRef.adUnitId = undefined;
```
Will be fixed in the next submission

agregorio-improve commented on improvedigitaladserverjsclient.js:
```diff
@@ -0,0 +1,230 @@
+'use strict';
```
In the next submission, all code will be merged into one file.

In terms of the file, this is our standard JS ad server client which contains code paths for multiple use cases and can be incorporated into any number of adapters.  As such, there will always be code paths that are not followed in this particular use case.

agregorio-improve commented on improvedigitaladserverjsclient.js:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    var returnIdMappings = true;
+    if (requestParameters.returnObjType === this.CONSTANTS.RETURN_OBJ_TYPE.PREBID) {
```
This is our standard JS ad server client which contains code paths for multiple use cases and can be incorporated into any number of adapters. As such, there will always be code paths that are not followed in this particular use case.  Either way, references to Prebid have been removed.

agregorio-improve commented on improve-digital-htb-validator.js:
```diff
@@ -42,10 +54,52 @@ var partnerValidator = function (configs) {
+                            currency: {
+                                type: 'string',
+                                optional: true,
+                                pattern: /^USD|EUR|GBP|AUD|DKK|SEK|CZK|CHF|NOK$/
```
Our protocol sets the currency at the impression level.  Although, there is no current business use case for multiple currencies in the same request, this protocol allows for maximum flexibility.

agregorio-improve commented:
Hi,

I have addressed all of your comments above and made the requested changes.  Please let me know if there are further issues/questions.

Best Regards,
Alfredo

agregorio-improve commented on improve-digital-htb.js:
```diff
@@ -511,4 +563,228 @@ function ImproveDigitalHtb(configs) {
+    }
+    var errors = null;
+
+    var baseUrl = (requestParameters.secure === 1 ? 'https' : 'http') + '://' + this.CONSTANTS.AD_SERVER_BASE_URL + '/' + this.CONSTANTS.END_POINT + '?' + this.CONSTANTS.AD_SERVER_URL_PARAM;
```
As explained previously, we have common code that we can slot into different header bidding implementations and this is one example.  We pass in a parameter called secure that can be set to CONSTANTS.SECURE or CONSTANTS.STANDARD.  We use Browser.getProtocol() to decide which one to pass in.  See : https://github.com/indexexchange/improve-digital-adapter/blob/5bb39acce4810c9c3d2c187bdbc6b63fc69c00cc/improve-digital-htb.js#L177

agregorio-improve commented on improve-digital-htb-validator.js:
```diff
@@ -42,10 +54,52 @@ var partnerValidator = function (configs) {
+                                }
+                            },
+                            publisherId: {
+                                type: 'number',
```
Extra validation has been added.

agregorio-improve commented:
I've just addressed the latest 2 comments.  Please take a look.

ix-certification commented on improve-digital-htb-validator.js:
```diff
@@ -19,6 +19,34 @@
+    mustIncludeOneOnly: function (schema, candidate) {
+        var mustIncludeOneOnly = schema.$mustIncludeOneOnly.params;
+        var occurrenceCounter = 0;
+        for (var index = 0; index < mustIncludeOneOnly.length; index++) {
```
When running the following:
```
let config = {
    "xSlots": {
        "1": {
            "placementId": 1053688
        },
        "2": {
            "publisherId": 1032,
            "placementKey": "data_team_test_hb_smoke_test"
        }
    },
    "mapping": {
        "slot1": [
            "1"
        ],
        "slot2": [
            "2"
        ]
    },
    "accountId": "8263",
    "timeout": 1000   
}
partnerValidator(config)
```
An error was thrown: 
```
TypeError: Cannot read property 'length' of undefined
```
at line 38.

Please make corresponding changes.

agregorio-improve replied:
I have fixed this now.  There was a typo.

ix-certification commented on spec/parseResponse.spec.js:
```diff
@@ -99,392 +101,55 @@ describe('parseResponse', function () {
+			})(i);
+		};
+		
+		after(function () {
```
The unit test failed as the function after is undefined. Please ensure all unit tests pass.

agregorio-improve replied:
All unit tests do pass.  Running the following:

npm install
mocha spec/*

gives the following output:

  generateRequestObj
    √ MRA - should return a correctly formatted object
    √ should correctly build a url
    √ MRA - should return a correctly formatted object
    √ should correctly build a url
    √ MRA - should return a correctly formatted object
    √ should correctly build a url
    √ MRA - should return a correctly formatted object
    √ should correctly build a url
    √ MRA - should return a correctly formatted object
    √ should correctly build a url
    √ MRA - should return a correctly formatted object
    √ should correctly build a url
    √ MRA - should return a correctly formatted object
    √ should correctly build a url
    √ MRA - should return a correctly formatted object
    √ should correctly build a url
    √ MRA - should return a correctly formatted object
    √ should correctly build a url
    √ MRA - should return a correctly formatted object
    √ should correctly build a url
    √ MRA - should return a correctly formatted object
    √ should correctly build a url
    √ MRA - should return a correctly formatted object
    √ should correctly build a url

  parseResponse
    should correctly parse bids:
      √ Check response for slot _2222222222
      √ Check response for slot _2222222223
      √ Check response for slot _2222222224
      √ Check response for slot _2222222225
      √ Check response for slot _2222222226

  Partner Profile
    √ should be configured correctly


  30 passing (155ms)

agregorio-improve replied:
Looks like calling mocha directly worked but calling npm test didn't.  Latest submission fixes this.
